examples/CustomCollectionView/Sample/MosaicCollectionViewLayout.m

```diff
@@ -0,0 +1,230 @@
+      
+      [_itemAttributes[section] addObject:attributes];
+      [_allAttributes addObject:attributes];
+    }
+    
+    NSUInteger columnIndex = [self _tallestColumnIndexInSection:section];
+    top = [_columnHeights[section][columnIndex] floatValue] - _interItemSpacing.bottom + _sectionInset.bottom;
+    
+    for (NSUInteger idx = 0; idx < [_columnHeights[section] count]; idx++) {
+      _columnHeights[section][idx] = @(top);
+    }
+  }
+}
+
+- (NSArray *)layoutAttributesForElementsInRect:(CGRect)rect
+{
+  NSMutableArray *includedAttributes = [NSMutableArray array];
+  // Slow search for small batches
+  for (UICollectionViewLayoutAttributes *attributes in _allAttributes) {
+    if (CGRectIntersectsRect(attributes.frame, rect)) {
+      [includedAttributes addObject:attributes];
+    }
+  }
+  return includedAttributes;
+}
+
+- (UICollectionViewLayoutAttributes *)layoutAttributesForItemAtIndexPath:(NSIndexPath *)indexPath
+{
+  if (indexPath.section >= _itemAttributes.count) {
+    return nil;
+  } else if (indexPath.item >= [_itemAttributes[indexPath.section] count]) {
+    return nil;
+  }
+  return _itemAttributes[indexPath.section][indexPath.item];
+}
+
+- (UICollectionViewLayoutAttributes *)layoutAttributesForSupplementaryViewOfKind:(NSString *)elementKind atIndexPath:(NSIndexPath *)indexPath
+{
+  if ([elementKind isEqualToString:UICollectionElementKindSectionHeader]) {
+    return _headerAttributes[@(indexPath.section)];
+  }
+  return nil;
+}
+
+- (BOOL)shouldInvalidateLayoutForBoundsChange:(CGRect)newBounds
+{
+  if (!CGRectEqualToRect(self.collectionView.bounds, newBounds)) {
+    return YES;
+  }
+  return NO;
+}
+
+- (CGFloat)_widthForSection:(NSUInteger)section
+{
+  return self.collectionView.bounds.size.width - _sectionInset.left - _sectionInset.right;
+}
+
+- (CGFloat)_columnWidthForSection:(NSUInteger)section
+{
+  return ([self _widthForSection:section] - ((_numberOfColumns - 1) * _columnSpacing)) / _numberOfColumns;
+}
+
+- (CGSize)_itemSizeAtIndexPath:(NSIndexPath *)indexPath
+{
+  CGSize size = CGSizeMake([self _columnWidthForSection:indexPath.section], 0);
+  CGSize originalSize = [[self _delegate] collectionView:self.collectionView layout:self originalItemSizeAtIndexPath:indexPath];
+  if (originalSize.height > 0 && originalSize.width > 0) {
+    size.height = originalSize.height / originalSize.width * size.width;
+  }
+  return size;
+}
+
+- (CGSize)_headerSizeForSection:(NSUInteger)section
+{
+  return CGSizeMake([self _widthForSection:section], _headerHeight);
+}
+
+- (CGSize)collectionViewContentSize
+{
+  CGFloat height = [[[_columnHeights lastObject] firstObject] floatValue];
+  return CGSizeMake(self.collectionView.bounds.size.width, height);
+}
+
+- (NSUInteger)_tallestColumnIndexInSection:(NSUInteger)section
+{
+  __block NSUInteger index = 0;
+  __block CGFloat tallestHeight = 0;
+  [_columnHeights[section] enumerateObjectsUsingBlock:^(NSNumber *height, NSUInteger idx, BOOL *stop) {
```

Comment:
Agreed, but I would like to optimize this layout for readability since it's just an example. Can use fast enumeration here, but `enumerateObjectsUsingBlock` is just so damn pretty :smile: